default/AI/ShipDesignAI.py

```diff
@@ -0,0 +1,1571 @@
+
+    def __init__(self):
+        """Make sure to call this constructor in each subclass."""
+        self.species = None         # name of the piloting species (string)
+        self.hull = None            # hull object (not hullname!)
+        self.partnames = []         # list of partnames (string)
+        self.parts = []             # list of actual part objects
+        self.attacks = {}           # {damage:count}
+        self.structure = 0
+        self.shields = 0
+        self.fuel = 0
+        self.speed = 0
+        self.stealth = 0
+        self.detection = 0
+        self.troops = 0
+        self.colonisation = -1      # -1 since 0 indicates an outpost (capacity = 0)
+        self.production_cost = 9999
+        self.production_time = 1
+        self.pid = -1               # planetID for checks on production cost if not LocationInvariant.
+        self.additional_specifications = AdditionalSpecifications()
+
+    def evaluate(self):
+        """ Return a rating for the design.
+
+        First, check if the minimum requirements are met. If so, return the _rating_function().
+        Otherwise, return a large negative number scaling with distance to the requirements.
+
+        :returns: float - rating of the current part/hull combo
+        """
+        self.update_stats()
+        rating = 0
+
+        # If we do not meet the requirements, we want to return a negative rating.
+        # However, we also need to make sure, that the closer we are to requirements,
+        # the better our rating is so the optimizing heuristic finds "the right way".
+        if self.fuel < self.additional_specifications.minimum_fuel:
+            rating += -1000 * (self.additional_specifications.minimum_fuel - self.fuel)
+        if self.speed < self.additional_specifications.minimum_speed:
+            rating += -1000 * (self.additional_specifications.minimum_speed - self.speed)
+        if self.structure < self.additional_specifications.minimum_structure:
+            rating += -1000 * (self.additional_specifications.minimum_structure - self.structure)
+        if rating < 0:
+            return rating
+        else:
+            return self._rating_function()
+
+    def _rating_function(self):
+        """Rate the design according to current hull/part combo.
+
+        :returns: float - rating
+        """
+        print_error("WARNING: Rating function not overloaded for class %s!" % self.__class__.__name__)
+        return -9999
+
+    def _set_stats_to_default(self):
+        """Set stats to default.
+
+        Call this if design is invalid to avoid miscalculation of ratings."""
+        self.structure = 0
+        self.attacks.clear()
+        self.shields = 0
+        self.fuel = 0.0001
+        self.speed = 0.0001
+        self.stealth = 0.0001
+        self.detection = 0.0001
+        self.troops = 0
+        self.colonisation = -1
+        self.production_cost = 9999999
```

Comment:
Should it be called at init? 
Should be `self.production_cost` set same as in init?  `self.production_cost = 9999`